File: app/src/main/java/org/hesab/app/AddTransactionActivity.java
package org.hesab.app;

import android.app.DatePickerDialog;
import android.os.Bundle;
import android.widget.*;
import androidx.appcompat.app.AppCompatActivity;
import java.text.DecimalFormat;
import java.util.Calendar;

public class AddTransactionActivity extends AppCompatActivity {

    private EditText etDate, etAmount, etCategory, etDescription;
    private Spinner spBank;
    private RadioButton rbIncome, rbExpense;
    private Button btnSave;
    private TransactionDao dao;
    private Transaction editTransaction;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_transaction);

        dao = TransactionDatabase.getInstance(this).transactionDao();

        etDate = findViewById(R.id.etDate);
        etAmount = findViewById(R.id.etAmount);
        etCategory = findViewById(R.id.etCategory);
        etDescription = findViewById(R.id.etDescription);
        spBank = findViewById(R.id.spBank);
        rbIncome = findViewById(R.id.rbIncome);
        rbExpense = findViewById(R.id.rbExpense);
        btnSave = findViewById(R.id.btnSave);

        etDate.setOnClickListener(v -> showDatePicker());
        etAmount.addTextChangedListener(new NumberTextWatcher(etAmount));

        int editId = getIntent().getIntExtra("editTransactionId", -1);
        if (editId != -1) {
            for (Transaction t : dao.getAll()) {
                if (t.id == editId) {
                    editTransaction = t;
                    etDate.setText(t.date);
                    etAmount.setText(String.valueOf(t.amount));
                    etCategory.setText(t.category);
                    etDescription.setText(t.description);
                    rbIncome.setChecked(t.isIncome);
                    rbExpense.setChecked(!t.isIncome);
                    break;
                }
            }
        }

        btnSave.setOnClickListener(v -> saveTransaction());
    }

    private void saveTransaction() {
        String date = etDate.getText().toString();
        long amount = Long.parseLong(etAmount.getText().toString().replace(",", ""));
        String category = etCategory.getText().toString();
        String description = etDescription.getText().toString();
        boolean isIncome = rbIncome.isChecked();
        String bank = spBank.getSelectedItem().toString();

        if (editTransaction != null) {
            editTransaction.date = date;
            editTransaction.amount = amount;
            editTransaction.category = category;
            editTransaction.description = description;
            editTransaction.isIncome = isIncome;
            editTransaction.bankName = bank;
            dao.update(editTransaction);
        } else {
            int orderIndex = dao.getAll().size();
            Transaction t = new Transaction(date, amount, category, description, isIncome, bank, orderIndex);
            dao.insert(t);
        }

        finish();
    }

    private void showDatePicker() {
        Calendar c = Calendar.getInstance();
        DatePickerDialog dpd = new DatePickerDialog(this,
                (view, year, month, day) -> {
                    PersianDate pDate = new PersianDate();
                    pDate.setGrgDate(year, month + 1, day);
                    PersianDateFormat pdForm = new PersianDateFormat("Y/m/d");
                    etDate.setText(pdForm.format(pDate));
                },
                c.get(Calendar.YEAR), c.get(Calendar.MONTH), c.get(Calendar.DAY_OF_MONTH));
        dpd.show();
    }
}

File: app/src/main/java/org/hesab/app/AddTransactionActivity.kt
package org.hesab.app

import android.app.DatePickerDialog
import android.os.Bundle
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.*

class AddTransactionActivity : AppCompatActivity() {

    private lateinit var db: AppDatabase
    private val dateFormat = SimpleDateFormat("yyyy/MM/dd", Locale.US)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_transaction)

        db = AppDatabase.getInstance(this)

        val radioIncome = findViewById<RadioButton>(R.id.radioIncome)
        val radioExpense = findViewById<RadioButton>(R.id.radioExpense)
        val etDate = findViewById<EditText>(R.id.etDate)
        val btnPickDate = findViewById<ImageButton>(R.id.btnPickDate)
        val etAmount = findViewById<EditText>(R.id.etAmount)
        val etCategory = findViewById<EditText>(R.id.etCategory)
        val etNote = findViewById<EditText>(R.id.etNote)
        val btnSave = findViewById<Button>(R.id.btnSave)
        val btnCancel = findViewById<Button>(R.id.btnCancel)

        // تاریخ امروز پیش‌فرض
        val cal = Calendar.getInstance()
        etDate.setText(dateFormat.format(cal.time))

        btnPickDate.setOnClickListener {
            val now = Calendar.getInstance()
            DatePickerDialog(this, { _, y, m, d ->
                val picked = Calendar.getInstance()
                picked.set(y, m, d)
                etDate.setText(dateFormat.format(picked.time))
            }, now.get(Calendar.YEAR), now.get(Calendar.MONTH), now.get(Calendar.DAY_OF_MONTH)).show()
        }

        btnCancel.setOnClickListener { finish() }

        btnSave.setOnClickListener {
            val amountText = etAmount.text.toString().replace(",", "").trim()
            val amount = try { amountText.toLong() } catch (e: Exception) { 0L }
            val category = etCategory.text.toString().ifBlank { "بدون عنوان" }
            val note = etNote.text.toString().ifBlank { "" }
            val isIncome = radioIncome.isChecked
            val dateString = etDate.text.toString().trim()

            // تاریخ را به timestamp تبدیل می‌کنیم (درصورت خطا، زمان فعلی را می‌گیرد)
            val dateMillis = try {
                dateFormat.parse(dateString)?.time ?: System.currentTimeMillis()
            } catch (e: Exception) {
                System.currentTimeMillis()
            }

            val txn = Transaction(
                date = dateMillis,
                amount = amount,
                category = category,
                note = note,
                isIncome = isIncome,
                orderIndex = 0
            )

            CoroutineScope(Dispatchers.IO).launch {
                db.transactionDao().insert(txn)
                runOnUiThread { finish() }
            }
        }
    }
}
